src/server/handlers.rs

```diff
@@ -762,3 +747,158 @@ impl<P: Publisher> Handler<P> {
+                return Some(format!(
+                    "Unable to serialize JSON config data for component {}: {e:?}",
+                    component.name
+                ));
+            }
+
+            if !component_details.insert(image_name.to_string()) {
+                return Some(format!(
+                    "Duplicate image reference in manifest: {}",
+                    image_name
+                ));
+            }
+
+            if !link_name_set.insert(link.to_string()) {
+                return Some(format!("Duplicate link name in manifest: {}", image_name));
+            }
+        }
+
+        // Actor validation : Actors should have a unique name and reference
+        if let Properties::Actor {
+            properties: ActorProperties { image: image_name },
+        } = &component.properties
+        {
+            if !component_details.insert(image_name.to_string()) {
+                return Some(format!(
+                    "Duplicate image reference in manifest: {}",
+                    image_name
+                ));
+            }
+        }
+
+        // Linkdef validation : A linkdef from a component should have a unique target and reference
+        let mut linkdef_set: HashSet<String> = HashSet::new();
+        if let Some(traits_vec) = &component.traits {
+            for trait_item in traits_vec.iter() {
+                if let Trait {
+                    // TODO : add trait type validation after custom types are done. See TraitProperty enum.
+                    properties:
+                        TraitProperty::Linkdef(LinkdefProperty {
+                            target: target_name,
+                            values: linkdef_values,
+                        }),
+                    ..
+                } = &trait_item
+                {
+                    if !linkdef_set.insert(target_name.to_string()) {
+                        return Some(format!(
+                            "Duplicate target for linkdef in manifest: {}",
+                            target_name
+                        ));
+                    }
+
+                    if !existing_linkdefs.is_empty()
+                        && linkdef_values.is_some()
+                        && !(existing_linkdefs
+                            .get(&(component.name.clone(), target_name.to_string()))
+                            .map(|x| x.to_owned())
+                            .eq(linkdef_values)
+                            || existing_linkdefs
+                                .get(&(target_name.to_string(), component.name.clone()))
+                                .map(|x| x.to_owned())
+                                .eq(linkdef_values))
+                    {
+                        return Some(format!(
+                            "Component {} and target {} have different linkdef values from earlier version of manifest",
+                            component.name, target_name
+                        ));
```

Comment:
This is actually an accepted scenario, but components and targets cannot have different values from _other_ existing deployed manifests 🤔 This is a really complicated check to do at validation time and I'm leaning towards leaving it out, we should pair on this later.

@thomastaylor312 do you have thoughts here? It's hard to check _every_ other manifest in the store in order to ensure link values are unique, especially because component/targets can have different names, image refs, etc

Reply:
Decided after syncing that this is too complicated to check in the scope of this PR, which is a really nice `validate_manifest` function that can work without querying _every_ other manifest.